packages/go_router/lib/src/go_route_information_parser.dart

```diff
@@ -4,20 +4,400 @@
+  /// with this parser.
+  ///
+  /// An assertion error will be thrown if this property set to true and the
+  /// [GoRouteInformationProvider] is in not in use.
+  ///
+  /// Defaults to false.
+  final bool debugRequireGoRouteInformationProvider;
+
+  final Map<String, String> _nameToPath = <String, String>{};
+
+  void _cacheNameToPath(String parentFullPath, List<GoRoute> childRoutes) {
+    for (final GoRoute route in childRoutes) {
+      final String fullPath = concatenatePaths(parentFullPath, route.path);
+
+      if (route.name != null) {
+        final String name = route.name!.toLowerCase();
+        assert(!_nameToPath.containsKey(name),
+            'duplication fullpaths for name "$name":${_nameToPath[name]}, $fullPath');
+        _nameToPath[name] = fullPath;
+      }
+
+      if (route.routes.isNotEmpty) {
+        _cacheNameToPath(fullPath, route.routes);
+      }
+    }
+  }
+
+  /// Looks up the url location by a [GoRoute]'s name.
+  String namedLocation(
+    String name, {
+    Map<String, String> params = const <String, String>{},
+    Map<String, String> queryParams = const <String, String>{},
+  }) {
+    assert(_nameToPath.containsKey(name), 'unknown route name: $name');
+    final String path = _nameToPath[name]!;
+    assert(() {
+      // Check that all required params are presented.
+      final List<String> paramNames = <String>[];
+      patternToRegExp(path, paramNames);
+      for (final String paramName in paramNames) {
+        assert(params.containsKey(paramName),
+            'missing param "$paramName" for $path');
+      }
+
+      // Check that there are no extra params
+      for (final String key in params.keys) {
+        assert(paramNames.contains(key), 'unknown param "$key" for $path');
+      }
+      return true;
+    }());
+    final Map<String, String> encodedParams = <String, String>{
+      for (final MapEntry<String, String> param in params.entries)
+        param.key: Uri.encodeComponent(param.value)
+    };
+    final String location = patternToPath(path, encodedParams);
+    return Uri(path: location, queryParameters: queryParams).toString();
+  }
+
+  /// Concatenates two paths.
+  ///
+  /// e.g: pathA = /a, pathB = c/d,  concatenatePaths(pathA, pathB) = /a/c/d.
+  static String concatenatePaths(String parentPath, String childPath) {
+    // at the root, just return the path
+    if (parentPath.isEmpty) {
+      assert(childPath.startsWith('/'));
+      assert(childPath == '/' || !childPath.endsWith('/'));
+      return childPath;
+    }
+
+    // not at the root, so append the parent path
+    assert(childPath.isNotEmpty);
+    assert(!childPath.startsWith('/'));
+    assert(!childPath.endsWith('/'));
+    return '${parentPath == '/' ? '' : parentPath}/$childPath';
+  }
+
   /// for use by the Router architecture as part of the RouteInformationParser
   @override
-  Future<Uri> parseRouteInformation(
+  Future<List<GoRouteMatch>> parseRouteInformation(
```

Comment:
It doesn't look like there are any tests for this method - I'll probably add some when I start working on  #99126.

Reply:
This method should be covered by a lot of test indirectly, any test that build the page from uri should do

Reply:
I can add it in this patch

Reply:
Not sure if it's helpful, but there's some test cases in this file that might be useful to consider: https://github.com/johnpryan/tree_router/blob/main/test/tree_test.dart